ledger/complete/wal/checkpoint_v6_reader.go

```diff
@@ -0,0 +1,340 @@
+package wal
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"path"
+
+	"github.com/onflow/flow-go/ledger/complete/mtrie/flattener"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/node"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+)
+
+// ReadCheckpointV6 reads checkpoint file from a main file and 17 file parts.
+// the main file stores:
+// 		1. version
+//		2. checksum of each part file (17 in total)
+// 		3. checksum of the main file itself
+// 	the first 16 files parts contain the trie nodes below the subtrieLevel
+//	the last part file contains the top level trie nodes above the subtrieLevel and all the trie root nodes.
+func ReadCheckpointV6(dir string, fileName string) ([]*trie.MTrie, error) {
+	// TODO: read the main file and check the version
+	// TODO: read the checksum of 17 file parts
+
+	headerPath := filePathHeader(dir, fileName)
+	subtrieChecksums, topTrieChecksum, err := readHeader(headerPath)
+	if err != nil {
+		return nil, fmt.Errorf("could not read header: %w", err)
+	}
+
+	subtrieNodes, err := readSubTriesConcurrently(dir, fileName, subtrieChecksums)
+	if err != nil {
+		return nil, fmt.Errorf("could not read subtrie from dir: %w", err)
+	}
+
+	// validate the subtrie node checksum
+	tries, err := readTopLevelTries(dir, fileName, subtrieNodes, topTrieChecksum)
+	if err != nil {
+		return nil, fmt.Errorf("could not read top level nodes or tries: %w", err)
+	}
+
+	return tries, nil
+}
+
+func filePathHeader(dir string, fileName string) string {
+	return path.Join(dir, fileName)
+}
+
+func filePathSubTries(dir string, fileName string, index int) (string, string, error) {
+	if index < 0 || index > 15 {
+		return "", "", fmt.Errorf("index must be between 1 to 16, but got %v", index)
+	}
+	subTrieFileName := fmt.Sprintf("%s.%03d", fileName, index)
+	return path.Join(dir, subTrieFileName), subTrieFileName, nil
+}
+
+func filePathTopTries(dir string, fileName string) (string, string) {
```

Comment:
>We can combine checkpoint of top level nodes with summary file. Subtrie files are independent of each other and can be processed in parallel. However, top level nodes file depends on subtrie files and it needs to be processed differently from subtrie files. I think it is confusing to have subtrie files at the same level as top level nodes file.

@fxamacker Let's discuss here about your comments regarding merging top level nodes to summary file.

One benefit of keeping them separate is that, you can [verify the checksums of subtrie files in the summary file](https://github.com/onflow/flow-go/pull/3319/files#diff-2ff6507ee260a1e37156b4cae4cc259e7688bbd0007993b20fa193824073105aR116) before using them to [verify the checksum of each sub trie part file](https://github.com/onflow/flow-go/pull/3319/files#diff-2ff6507ee260a1e37156b4cae4cc259e7688bbd0007993b20fa193824073105aR203). 

Otherwise, I would be using unverified data (subtrie checksums in the summary file) to verify other data (subtrile part file's checksum), which might cause the error message to be confusing. 

Of course, I could verify the summary file checksum first before using those subtrie checksums , but then I will have to read all top level trie nodes, however, since I won't use those top level trie nodes until after processing all subtrie nodes, I would either have to buffer those top level tries data or read them again. 

That's why I chose to have the top level tries nodes stored in a separate file.